pluginutil/pluginutil.go

```diff
@@ -0,0 +1,197 @@
+// or filesystems and builds an overall mapping of a plugin name to its
+// information. The desired plugin can then be sent to CreatePlugin to actually
+// instantiate it. If a plugin is specified by name multiple times in option,
+// the last one wins.
+func BuildPluginMap(opt ...Option) (map[string]*PluginInfo, error) {
+	opts, err := GetOpts(opt...)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing plugin options: %w", err)
+	}
+
+	if len(opts.withPluginSources) == 0 {
+		return nil, fmt.Errorf("no plugins available")
+	}
+
+	pluginMap := map[string]*PluginInfo{}
+	for _, sourceInfo := range opts.withPluginSources {
+		switch {
+		case sourceInfo.pluginFs != nil:
+			if opts.withPluginClientCreationFunc == nil {
+				return nil, fmt.Errorf("non-in-memory plugin found but no creation func provided")
+			}
+			dirs, err := fs.ReadDir(sourceInfo.pluginFs, ".")
+			if err != nil {
+				return nil, fmt.Errorf("error scanning plugins: %w", err)
+			}
+			// Store a match between the config type string and the expected plugin name
+			for _, entry := range dirs {
+				pluginType := strings.TrimSuffix(strings.TrimPrefix(entry.Name(), sourceInfo.pluginFsPrefix), ".gz")
+				if runtime.GOOS == "windows" {
+					pluginType = strings.TrimSuffix(pluginType, ".exe")
+				}
+				pluginMap[pluginType] = &PluginInfo{
+					ContainerFs:              sourceInfo.pluginFs,
+					Filename:                 entry.Name(),
+					PluginClientCreationFunc: opts.withPluginClientCreationFunc,
+				}
+			}
+		case sourceInfo.pluginMap != nil:
+			for k, creationFunc := range sourceInfo.pluginMap {
+				pluginMap[k] = &PluginInfo{InmemCreationFunc: creationFunc}
+			}
+		}
+	}
+
+	return pluginMap, nil
+}
+
+// CreatePlugin instantiates a given plugin either via an in-memory function or
+// by executing a go-plugin plugin. The interface returned will either be a
+// *<go-plugin>.Client or the value returned from an in-memory function. A type
+// switch should be used by the calling code to determine this, and the
+// appropriate service should be Dispensed if what is returned is a go-plugin
+// plugin.
+//
+// If the WithSecureConfig option is passed, this will be round-tripped into the
+// PluginClientCreationFunction from the given *PluginInfo, where it can be sent
+// into the go-plugin client configuration.
+func CreatePlugin(plugin *PluginInfo, opt ...Option) (interface{}, func() error, error) {
+	switch {
+	case plugin.InmemCreationFunc != nil:
+		raw, err := plugin.InmemCreationFunc()
+		return raw, nil, err
+
+	case plugin.Filename == "" || plugin.PluginClientCreationFunc == nil:
+		return nil, nil, fmt.Errorf("no inmem creation func and either filename or plugin creation func not provided")
+	}
+
+	opts, err := GetOpts(opt...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing plugin options: %w", err)
+	}
+
+	// Open and basic validation
+	file, err := plugin.ContainerFs.Open(plugin.Filename)
```

Comment:
should we check if plugin.ContainerFs is nil?

Reply:
We should!